### PR TITLE
Fix MPI makefile race condition for boost

### LINKED ISSUE #1999: MPI tests rebuild when nothing has changed; also breaks on rebuild
## Description

When building tests with MPI, boost is rebuilt every time make is called with the test target. When boost is rebuilt for the second time, it will break the ability to link against the tests.

Looking at `make/libraries`, it looks like it expects a few `.so` files to exist, but they aren't generated. This will trigger a rebuild each time.

## Example

1. First add MPI settings to `make/local`.
2. Run `./runTests.py test/unit/math/prim/functor/mpi_parallel_call_test.cpp`; this should work
3. Run `./runTests.py test/unit/math/prim/functor/mpi_parallel_call_test.cpp` again; this will fail.

## Expected Output

On the second call to`./runTests.py test/unit/math/prim/functor/mpi_parallel_call_test.cpp` this should just run the test, but not rebuild it.

#### Current Version:
v3.3.0

### LINKED ISSUE #2012: Windows: build fails if make is located in a folder with spaces
If a user on Windows installs RTools in a folder with spaces, mingw32-make fails to build the math libs. Brackets also cause issues.

By default Rtools is not installed in such a path, but its not that uncommon that a user installs in

C:/Program Files(x86)/Rtools...
or
C:/Program Files/Rtools...

A similar issue pops up if you set the MAKE variable as "make -j4" or something else, see https://github.com/stan-dev/cmdstanr/issues/278

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

Fixes #1999 and fixes #2012 

See https://github.com/stan-dev/math/issues/1999#issuecomment-672659549 for a description of the error

In order to verify run

Run 
`make lib/boost_1.72.0/stage/lib/libboost_mpi.so  lib/boost_1.72.0/stage/lib/libboost_serialization.so -j2
`on linux or 
`make lib/boost_1.72.0/stage/lib/libboost_mpi.dylib  lib/boost_1.72.0/stage/lib/libboost_serialization.dylib -j2
`on a Mac.

## Tests

/

## Side Effects

/

## Release notes

Fixed issue building the Math libraries on Windows if the C++ toolchain was installed in a folder with spaces.

## Checklist

- [x] Math issue #(issue number)

- [x] Copyright holder: Rok Češnovar

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
With develop running

make lib/boost_1.72.0/stage/lib/libboost_mpi.so lib/boost_1.72.0/stage/lib/libboost_serialization.so -j2 

I see

cd lib/boost_1.72.0; ./bootstrap.sh
cd lib/boost_1.72.0; ./bootstrap.sh
and then everything is duplicated.

On this branch there is no duplication. Might be good to double-check on a mac.

This issue is annoying as it produces a bunch of false positives on Jenkins tests.

**Comment by rok-cesnovar** [comment]
I also added a fix for #2012 in this PR so we don't need a separate PR for that.

**Comment by bbbales2** [comment]
This fixes problems with:

```make lib/boost_1.72.0/stage/lib/libboost_mpi.so lib/boost_1.72.0/stage/lib/libboost_serialization.so -j2```

running the boost build twice. I wasn't able to confirm #1999 though (it just works in develop for me).

Hopping over to Windows to check the other thing.

**Comment by rok-cesnovar** [comment]
@bbbales2 I reponed 1999 so that we have an issue for the 

make lib/boost_1.72.0/stage/lib/libboost_mpi.so lib/boost_1.72.0/stage/lib/libboost_serialization.so -j2

problem. The basic problem of 1999 is fixed on develop.

Thanks for testing and the review for RTools!

**Comment by bbbales2** [comment]
Things don't seem to build in Windows and I'm getting path errors:
```
$ mingw32-make test/unit/math/rev/functor/ode_adams_rev_test
g++ -std=c++1y -m64 -D_REENTRANT -Wall -Wno-unused-function -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-variable -Wno-sign-compare -Wno-unused-local-typedefs -Wno-int-in-bool-context -Wno-attributes -Wno-ignored-attributes      -I lib/tbb_2019_U8/include -O3  -I . -I lib/eigen_3.3.7 -I lib/boost_1.72.0 -I lib/sundials_5.2.0/include -I lib/benchmark_1.5.1/googletest/googletest/include -I lib/benchmark_1.5.1/googletest/googletest   -D_USE_MATH_DEFINES  -DBOOST_DISABLE_ASSERTS             -Wl,-L,"C:/Users/admin/math/lib/tbb" -Wl,-rpath,"C:/Users/admin/math/lib/tbb"  test/unit/math/rev/functor/ode_adams_rev_test.cpp   -static-libgcc -static-libstdc++       -o test/unit/math/rev/functor/ode_adams_rev_test
C:/Program Files (x86)/rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find C:/Program: No such file or directory
C:/Program Files (x86)/rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find Files: No such file or directory
C:/Program Files (x86)/rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find (x86)/rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/lib/../lib/default-manifest.o: No such file or directory
collect2.exe: error: ld returned 1 exit status
mingw32-make: *** No rule to make target 'test/unit/math/rev/functor/ode_adams_rev_test'.  Stop.
```

Is my test build appropriate?

**Comment by rok-cesnovar** [comment]
I merely tried 

python runTests.py test/unit/..

I think that should also be with a .exe at the end:
mingw32-make test/unit/math/rev/functor/ode_adams_rev_test.exe

**Comment by bbbales2** [comment]
I got a python installed so I could do the run tests and I'm getting the same thing:

rtools is in "C:\Program Files(x86)\"

```
$ python runTests.py test/unit/math/rev/functor/ode_adams_rev_test.cpp
------------------------------------------------------------
mingw32-make -j1 test/unit/math/rev/functor/ode_adams_rev_test.exe
g++ -std=c++1y -m64 -D_REENTRANT -Wall -Wno-unused-function -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-variable -Wno-sign-compare -Wno-unused-local-typedefs -Wno-int-in-bool-context -Wno-attributes -Wno-ignored-attributes      -I lib/tbb_2019_U8/include -O3  -I . -I lib/eigen_3.3.7 -I lib/boost_1.72.0 -I lib/sundials_5.2.0/include -I lib/benchmark_1.5.1/googletest/googletest/include -I lib/benchmark_1.5.1/googletest/googletest -I lib/benchmark_1.5.1/googletest/googletest/include -I lib/benchmark_1.5.1/googletest/googletest    -D_USE_MATH_DEFINES  -DBOOST_DISABLE_ASSERTS        -c -o test/unit/math/rev/functor/ode_adams_rev_test.o test/unit/math/rev/functor/ode_adams_rev_test.cpp
g++ -std=c++1y -m64 -D_REENTRANT -Wall -Wno-unused-function -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-variable -Wno-sign-compare -Wno-unused-local-typedefs -Wno-int-in-bool-context -Wno-attributes -Wno-ignored-attributes      -I lib/tbb_2019_U8/include -O3  -I . -I lib/eigen_3.3.7 -I lib/boost_1.72.0 -I lib/sundials_5.2.0/include -I lib/benchmark_1.5.1/googletest/googletest/include -I lib/benchmark_1.5.1/googletest/googletest   -D_USE_MATH_DEFINES  -DBOOST_DISABLE_ASSERTS             -Wl,-L,"C:/Users/admin/math/lib/tbb" -Wl,-rpath,"C:/Users/admin/math/lib/tbb"  test/unit/math/rev/functor/ode_adams_rev_test.o lib/benchmark_1.5.1/googletest/googletest/src/gtest_main.cc lib/benchmark_1.5.1/googletest/googletest/src/gtest-all.o lib/tbb/tbb.dll lib/sundials_5.2.0/lib/libsundials_nvecserial.a lib/sundials_5.2.0/lib/libsundials_cvodes.a lib/sundials_5.2.0/lib/libsundials_idas.a lib/sundials_5.2.0/lib/libsundials_kinsol.a  -static-libgcc -static-libstdc++       -o test/unit/math/rev/functor/ode_adams_rev_test.exe
C:/Program Files (x86)/rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find C:/Program: No such file or directory
C:/Program Files (x86)/rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find Files: No such file or directory
C:/Program Files (x86)/rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find (x86)/rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/lib/../lib/default-manifest.o: No such file or directory
collect2.exe: error: ld returned 1 exit status
mingw32-make: *** [make/tests:13: test/unit/math/rev/functor/ode_adams_rev_test.exe] Error 1
mingw32-make -j1 test/unit/math/rev/functor/ode_adams_rev_test.exe failed
```

**Comment by bbbales2** [comment]
I checked I'm on the `bugfix/mpi-makefile` branch. I also tried ` python runTests.py test/unit/math/prim/fun/add_test.cpp` to the same end.

**Comment by rok-cesnovar** [comment]
Ok, thanks for digging into this. Will spin a windows instance and have another go. I placed it in the same spot (Program files x86) , though it may have been RTools 3.5 as I was looking into the other issues.

**Comment by rok-cesnovar** [comment]
Dug deeper here and what I found was basically that if the toolchain is installed in a path with spaces and brackets, it won't work. I tried compiling and linking manually and it fails. Will put in a nicer error message, to at least give some suggestion to what is wrong.

**Comment by bbbales2** [comment]
Oh wow that's annoying. So like just spaces are fine but spaces and brackets are a no-go? Or it's a no-go period?

**Comment by bbbales2** [comment]
Annoying in the annoying it doesn't work -- a warning is a fine solution

**Comment by rok-cesnovar** [comment]
Essentially spaces or brackets are a no go for the toolchain install path. 

That's annoying, but at least not as annoying as if we had problems with the location of cmdstan/math itself. That would be more limiting to the user.

I should have figured that given that even R devs haven't figured it out yet (if you install RTools in a path with spaces you cant any install any R package from source), that we are probably facing the same problem that we won't be able to fix.

**Comment by bbbales2** [comment]
Cool beans sounds good.

**Comment by rok-cesnovar** [comment]
Ready now.

File: make/libraries [hunk 1]
```diff
@@ -131,7 +137,18 @@ ifeq (Linux, $(OS))
...
+# we error, as those characters cause issues when building.
+ifeq (Windows_NT, $(OS))
+MAKE_ESCAPED := ${subst ),,${subst (,,$(lastword $(MAKE))}}
+else
+MAKE_ESCAPED := $(MAKE)
+endif
+
 $(TBB_BIN)/tbb-make-check:
+ifneq ($(MAKE),$(MAKE_ESCAPED))
+$(error '$nError:$n$nThe toolchain is installed in a path with spaces or brackets.$nPlease reinstall the toolchain.$n$n')
```
Comment (bbbales2) [inline make/libraries:150]: Add rtools somewhere -- this is only happening in Windows right and the only toolchain people are using is rtools?

Comment (rok-cesnovar) [reply make/libraries:150]: Good call. And yes to both questions.

**Comment by rok-cesnovar** [comment]
> the error pops up when I do make but not when I call this from python. Any ideas what that is

Let me double-check. This error will generally pop up when building TBB, maybe a dependency thing.

> Sorry for the slow review,

No problem, not super urgent turns out now that gelman linux machine is back up.

File: make/libraries (hunk 1 above)
Comment (rok-cesnovar) [reply make/libraries]: New message is: 

The RTools toolchain is installed in a path with spaces or brackets.
Please reinstall the toolchain.

**Comment by rok-cesnovar** [comment]
> the error pops up when I do make but not when I call this from python. Any ideas what that is

The problem is that if run via make then $(MAKE) is the full path to mingw32-make. If run through python its only "mingw32-make" and we cant detect spaces and brackets. So I added a check to the python script as well. The script has a bunch of changes because I ran a python code formatted we usually use to format this script, but haven't in a while.

**Comment by rok-cesnovar** [comment]
This usually happens if you do not have Path to lib/tbb in PATH.

**Comment by bbbales2** [comment]
How do I add that to my PATH in Windows? This is the error I get:

```
$ test/unit/math/rev/functor/ode_adams_rev_test.exe --gtest_output="xml:test/unit/math/rev/functor/ode_adams_rev_test.xml"
C:/Users/admin/math/test/unit/math/rev/functor/ode_adams_rev_test.exe: error while loading shared libraries: ?: cannot open shared object file: No such file or directory
```

I added math/lib/tbb to my system PATH already.

**Comment by bbbales2** [comment]
Is there like an equivalent of `LD_LIBRARY_PATH` I could set in rtools?

**Comment by rok-cesnovar** [comment]
So you have `C:/Users/admin/math/lib/tbb` in PATH and you have restarted the shell?

**Comment by bbbales2** [comment]
Aye

**Comment by rok-cesnovar** [comment]
> Is there like an equivalent of `LD_LIBRARY_PATH` I could set in rtools?

You dont have to do anything else there. 

Can you try some other model, that does not require cvodes?

**Comment by bbbales2** [comment]
Same error:

```
$ test/unit/math/rev/fun/multiply_test.exe
C:/Users/admin/math/test/unit/math/rev/fun/multiply_test.exe: error while loading shared libraries: ?: cannot open shared object file: No such file or directory
```

I did a make clean-all and rebuilt things too.

**Comment by bbbales2** [comment]
There's a solid chance my Windows is just configured badly. The make seems to build fine. If the windows stuff passes here and it runs for you, I'm guessing there's just something screwed up in my Windows config, though I don't know what it is.

**Comment by rok-cesnovar** [comment]
You have RTools 4.0? What does g++ --version and mingw32-make --version return?

**Comment by bbbales2** [comment]
g++ 8.3.0
mingw32-make 4.2.1

**Comment by rok-cesnovar** [comment]
What happens if you just try to run the test directly? 

./test/unit/math/rev/fun/multiply_test.exe

Not via python runTests

**Comment by bbbales2** [comment]
This is what ldd says:

```
$ ldd test/unit/math/rev/fun/multiply_test.exe
        ntdll.dll => /c/WINDOWS/SYSTEM32/ntdll.dll (0x7ffc6a9c0000)
        KERNEL32.DLL => /c/WINDOWS/System32/KERNEL32.DLL (0x7ffc6a240000)
        KERNELBASE.dll => /c/WINDOWS/System32/KERNELBASE.dll (0x7ffc67a90000)
        msvcrt.dll => /c/WINDOWS/System32/msvcrt.dll (0x7ffc6a140000)
        USER32.dll => /c/WINDOWS/System32/USER32.dll (0x7ffc697e0000)
        win32u.dll => /c/WINDOWS/System32/win32u.dll (0x7ffc68790000)
        GDI32.dll => /c/WINDOWS/System32/GDI32.dll (0x7ffc68a70000)
        gdi32full.dll => /c/WINDOWS/System32/gdi32full.dll (0x7ffc688b0000)
        msvcp_win.dll => /c/WINDOWS/System32/msvcp_win.dll (0x7ffc687c0000)
        ucrtbase.dll => /c/WINDOWS/System32/ucrtbase.dll (0x7ffc67da0000)
```

**Comment by bbbales2** [comment]
Running directly fails too

**Comment by rok-cesnovar** [comment]
Can you try just copying the tbb.dll next to the `test/unit/math/rev/fun/multiply_test.exe` and then run?

**Comment by bbbales2** [comment]
Weird. I just booted up Rstudio and tried this in that console and it worked. So something about my rtools is a bit messed up that Rstudio fixes? Who knows.

```
$ test/unit/math/rev/fun/multiply_test.exe
Running main() from lib/benchmark_1.5.1/googletest/googletest/src/gtest_main.cc
?[0;32m[==========] ?[mRunning 0 tests from 0 test suites.
?[0;32m[==========] ?[m0 tests from 0 test suites ran. (7 ms total)
?[0;32m[  PASSED  ] ?[m0 tests.
```

**Comment by rok-cesnovar** [comment]
Do you work in the powershell or the old cmd? This seems bizarre.

**Comment by bbbales2** [comment]
In the Rtools 64bit shell.

I just checked PATH in the Rtools shell and it didn't have TBB. I added the path and it worked. So maybe my Rtools wasn't loading my system PATH?

**Comment by bbbales2** [comment]
(Also moving tbb to be in the folder I'm running the exe from worked)

**Comment by bbbales2** [comment]
Yuck not being familiar with OSes is confusing!

**Comment by rok-cesnovar** [comment]
Oh, the rtools shell. I never use that. I either use powershell or cmd. You can usually at least google issues for those, as opposed to these less common shells.

Rtools shell does not pick up the system PATH as its essentialy a "virtual environment". So you have to add it each time you start it.

**Pull request merged**